File: scripts/solver/solve.py
#!/usr/bin/env python3
"""
MILP Solver for Pen the Pet

Uses Mixed Integer Linear Programming (PuLP + CBC) to find the optimal
wall placement that maximizes the enclosed area around the pet's home tile.

The problem is formulated as a vertex-cut problem with flow-based connectivity:
- Binary variables for wall placement and pen membership
- Network flow ensures the pen is a connected region containing home
- Vertex-cut constraints ensure the pen boundary is entirely walls/water
- Boundary tiles are excluded from the pen (pet would escape)

Usage:
    echo '{"map": [...], "maxWalls": 7}' | python3 solve.py
    python3 solve.py --file input.json

Input JSON:
    map: 2D array of tile types ("grass", "water", "home")
    maxWalls: Maximum number of walls that can be placed

Output JSON:
    goalArea: Maximum enclosed area achievable
    optimalWallCount: Number of walls used in optimal solution
    optimalSolution: Array of [row, col] wall positions
    feasible: Whether a valid penning solution exists
"""

import sys
import json
import os
import subprocess
import argparse
from pulp import LpProblem, LpMaximize, LpVariable, lpSum, value, PULP_CBC_CMD


def load_tile_data():
    """
    Load tile data from js/tileData.js (single source of truth).

    Uses Node.js to require() the JS module and output tile data as JSON.
    This avoids duplicating tile definitions — the JS file is the only place
    where score, wallPlaceable, blocksMovement, etc. are defined.

    Returns:
        dict mapping tile name (str) → dict of all tile properties
    """
    js_path = os.path.abspath(
        os.path.join(os.path.dirname(__file__), '..', '..', 'js', 'tileData.js')
    )
    # Use JSON to safely escape the path in the Node.js expression
    safe_path = json.dumps(js_path.replace('\\', '/'))
    node_script = (
        f'const t = require({safe_path}); '
        'const d = {}; '
        'for (const [k, v] of Object.entries(t.TILE_DATA)) { '
        '  const e = {...v}; delete e.ariaLabel; d[k] = e; '
        '} '
        'console.log(JSON.stringify(d));'
    )
    try:
        result = subprocess.run(
            ['node', '-e', node_script],
            capture_output=True, text=True, check=True,
        )
    except FileNotFoundError:
        raise RuntimeError(
            'Node.js is required to load tile data from js/tileData.js. '
            'Please install Node.js (v24+).'
        )
    except subprocess.CalledProcessError as exc:
        raise RuntimeError(
            f'Failed to load tile data from {js_path}: {exc.stderr.strip()}'
        )
    return json.loads(result.stdout)


def load_tile_scores():
    """
    Load tile score values from js/tileData.js (single source of truth).

    Returns:
        dict mapping tile name (str) → score (int)
    """
    tile_data = load_tile_data()
    return {name: data['score'] for name, data in tile_data.items()}


def load_tile_properties():
    """
    Load tile properties from js/tileData.js (single source of truth).

    Returns:
        dict mapping tile name (str) → dict with blocksMovement, wallPlaceable,
        wallTransformsTo (optional)
    """
    tile_data = load_tile_data()
    return {
        name: {
            'blocksMovement': data['blocksMovement'],
            'wallPlaceable': data['wallPlaceable'],
            'wallTransformsTo': data.get('wallTransformsTo'),
        }
        for name, data in tile_data.items()
    }


def solve_map(map_data, max_walls):
    """
    Solve for optimal wall placement using MILP.

    Supports fillable tiles (blocksMovement=True AND wallPlaceable=True, e.g. holes).
    When a wall is placed on a fillable tile, it becomes passable and scores like
    its wallTransformsTo tile (e.g. hole → filledHole with score=1).

    Args:
        map_data: 2D list of tile type strings ("grass", "water", "home", "star", "hole", etc.)
        max_walls: Maximum number of walls allowed

    Returns:
        dict with goalArea, optimalWallCount, optimalSolution, feasible
    """
    tile_scores = load_tile_scores()
    tile_props = load_tile_properties()
    rows = len(map_data)
    cols = len(map_data[0])

    # Classify tile positions.
    # "graph_tiles" includes all non-blocking tiles PLUS fillable tiles (which
    # block when empty but become passable when filled).
    graph_tiles = []       # All tiles that participate in the MILP graph
    wall_placeable = []    # Tiles where walls can be placed (includes fillable)
    fillable_set = set()   # Fillable tile positions (blocksMovement AND wallPlaceable)
    home = None
    boundary = set()
    tile_score_map = {}    # (r,c) → score when in pen

    tile_set = set()
    for r in range(rows):
        for c in range(cols):
            tile = map_data[r][c]
            props = tile_props.get(tile, {})
            blocks = props.get('blocksMovement', False)
            placeable = props.get('wallPlaceable', False)
            is_fillable = blocks and placeable

            # Skip purely blocking tiles (water) that can't be filled
            if blocks and not placeable:
                continue

            graph_tiles.append((r, c))
            tile_set.add((r, c))

            if is_fillable:
                fillable_set.add((r, c))
                # Score when filled: use wallTransformsTo tile's score
                transformed = props.get('wallTransformsTo')
                tile_score_map[(r, c)] = tile_scores.get(transformed, 1) if transformed else 1
            else:
                tile_score_map[(r, c)] = tile_scores.get(tile, 1)

            if placeable:
                wall_placeable.append((r, c))
            if tile == 'home':
                home = (r, c)
            if r == 0 or r == rows - 1 or c == 0 or c == cols - 1:
                boundary.add((r, c))

    if home is None:
        return {"goalArea": 0, "optimalWallCount": 0, "optimalSolution": [], "feasible": False,
                "error": "No home tile found"}

    if home in boundary:
        return {"goalArea": 0, "optimalWallCount": 0, "optimalSolution": [], "feasible": False,
                "error": "Home is on boundary - cannot pen"}

    # Build directed edges between adjacent graph tiles
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    edges = []
    out_edges = {t: [] for t in graph_tiles}
    in_edges = {t: [] for t in graph_tiles}

    for (r, c) in graph_tiles:
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            if (nr, nc) in tile_set:
                edges.append(((r, c), (nr, nc)))
                out_edges[(r, c)].append((nr, nc))
                in_edges[(nr, nc)].append((r, c))

    n = len(graph_tiles)
    placeable_set = set(wall_placeable)

    # Create MILP problem
    prob = LpProblem("PenThePet", LpMaximize)

    # Variables: s[i] = 1 if tile i is in the pen
    s = {}
    for tile in graph_tiles:
        s[tile] = LpVariable(f"s_{tile[0]}_{tile[1]}", cat='Binary')

    # Variables: w[i] = 1 if wall placed on a wall-placeable tile i
    w = {}
    for tile in wall_placeable:
        w[tile] = LpVariable(f"w_{tile[0]}_{tile[1]}", cat='Binary')

    # Variables: f[(i,j)] = flow from tile i to tile j
    f = {}
    for (i, j) in edges:
        f[(i, j)] = LpVariable(f"f_{i[0]}_{i[1]}_{j[0]}_{j[1]}", lowBound=0)

    # Objective: maximize enclosed area using per-tile score values.
    # Tiny wall penalty as tiebreaker for minimum wall count.
    prob += lpSum(
        tile_score_map.get(tile, 1) * s[tile]
        for tile in graph_tiles
    ) - 0.0001 * lpSum(w[tile] for tile in wall_placeable)

    # Constraint 1: home is in the pen
    prob += s[home] == 1

    # Constraint 2: boundary tiles are NOT in the pen
    for tile in graph_tiles:
        if tile in boundary:
            prob += s[tile] == 0

    # Constraint 3a: Normal wall-placeable tiles — walled tiles are NOT in pen
    for tile in wall_placeable:
        if tile not in fillable_set:
            prob += s[tile] <= 1 - w[tile]

    # Constraint 3b: Fillable tiles — can only be in pen when filled (wall placed)
    # When w[tile]=0 (unfilled), s[tile]=0 (blocks movement, not in pen)
    # When w[tile]=1 (filled), s[tile] can be 0 or 1 (passable, may be in pen)
    for tile in fillable_set:
        prob += s[tile] <= w[tile]

    # Constraint 4: wall budget
    prob += lpSum(w[tile] for tile in wall_placeable) <= max_walls

    # Constraint 5: vertex cut — adjacent tiles on different sides need a barrier.
    # For normal tiles: barrier = w[tile] (wall placed = barrier)
    # For fillable tiles: barrier = 1 - w[tile] (unfilled = barrier, filled = passable)
    for (i, j) in edges:
        if i in placeable_set:
            barrier_i = (1 - w[i]) if i in fillable_set else w[i]
        else:
            barrier_i = 0
        if j in placeable_set:
            barrier_j = (1 - w[j]) if j in fillable_set else w[j]
        else:
            barrier_j = 0
        prob += s[i] - s[j] <= barrier_i + barrier_j

    # Constraint 6: flow conservation (pen connectivity via home)
    prob += (
        lpSum(f[(home, j)] for j in out_edges[home] if (home, j) in f) -
        lpSum(f[(j, home)] for j in in_edges[home] if (j, home) in f)
    ) == lpSum(s[tile] for tile in graph_tiles if tile != home)

    for tile in graph_tiles:
        if tile == home:
            continue
        inflow = lpSum(f[(j, tile)] for j in in_edges[tile] if (j, tile) in f)
        outflow = lpSum(f[(tile, j)] for j in out_edges[tile] if (tile, j) in f)
        prob += inflow - outflow == s[tile]

    # Constraint 7: flow capacity — flow only through pen tiles
    for (i, j) in edges:
        prob += f[(i, j)] <= n * s[i]
        prob += f[(i, j)] <= n * s[j]

    # Solve with CBC
    solver = PULP_CBC_CMD(msg=0, timeLimit=120)
    prob.solve(solver)

    if prob.status != 1:
        return {"goalArea": 0, "optimalWallCount": 0, "optimalSolution": [], "feasible": False,
                "error": f"Solver status: {prob.status}"}

    # Extract solution
    goal_area = int(round(value(prob.objective)))
    wall_positions = []
    for tile in wall_placeable:
        if value(w[tile]) is not None and value(w[tile]) > 0.5:
            wall_positions.append(list(tile))

    return {
        "goalArea": goal_area,
        "optimalWallCount": len(wall_positions),
        "optimalSolution": wall_positions,
        "feasible": True
    }


def main():
    parser = argparse.ArgumentParser(description='MILP Solver for Pen the Pet')
    parser.add_argument('--file', '-f', help='Input JSON file (default: read from stdin)')
    args = parser.parse_args()

    # Read input
    if args.file:
        with open(args.file, 'r') as fh:
            input_data = json.load(fh)
    else:
        input_data = json.load(sys.stdin)

    map_data = input_data['map']
    max_walls = input_data['maxWalls']

    # Solve
    result = solve_map(map_data, max_walls)

    # Output
    print(json.dumps(result))


if __name__ == '__main__':
    main()

File: scripts/solver/solve.js
/**
 * MILP Solver for Pen the Pet
 *
 * Uses Mixed Integer Linear Programming (GLPK via glpk.js) to find the optimal
 * wall placement that maximises the enclosed area around the pet's home tile.
 *
 * The problem is formulated as a vertex-cut problem with flow-based connectivity:
 * - Binary variables for wall placement (w) and pen membership (s)
 * - Network flow ensures the pen is a connected region containing home
 * - Vertex-cut constraints ensure the pen boundary is entirely walls/water
 * - Boundary tiles are excluded from the pen (pet would escape)
 *
 * Usage (CLI):
 *   echo '{"map": [...], "maxWalls": 7}' | node solve.js
 *   node solve.js --file input.json
 *
 * Input JSON:
 *   map: 2D array of tile type strings ("grass", "water", "home", ...)
 *   maxWalls: Maximum number of walls that can be placed
 *
 * Output JSON:
 *   goalArea: Maximum enclosed area achievable
 *   optimalWallCount: Number of walls used in optimal solution
 *   optimalSolution: Array of [row, col] wall positions
 *   feasible: Whether a valid penning solution exists
 */

'use strict';

const GLPK = require('glpk.js');
const { TILE_DATA } = require('../../js/tiles/tileData.js');

// ---------------------------------------------------------------------------
// Tile data helpers
// ---------------------------------------------------------------------------

function getTileScores() {
    const scores = {};
    for (const [name, data] of Object.entries(TILE_DATA)) {
        scores[name] = data.score;
    }
    return scores;
}

function getTileProperties() {
    const props = {};
    for (const [name, data] of Object.entries(TILE_DATA)) {
        props[name] = {
            blocksMovement: data.blocksMovement,
            wallPlaceable:  data.wallPlaceable,
            wallTransformsTo: data.wallTransformsTo,
        };
    }
    return props;
}

// ---------------------------------------------------------------------------
// Main solver
// ---------------------------------------------------------------------------

/**
 * Solve for optimal wall placement using MILP.
 *
 * Supports fillable tiles (blocksMovement=true AND wallPlaceable=true, e.g. holes).
 * When a wall is placed on a fillable tile, it becomes passable and scores like
 * its wallTransformsTo tile (e.g. hole → filledHole with score=1).
 *
 * @param {string[][]} mapData - 2D array of tile type strings
 * @param {number} maxWalls - Maximum number of walls allowed
 * @returns {{ goalArea, optimalWallCount, optimalSolution, feasible, error? }}
 */
function solveMap(mapData, maxWalls) {
    const glpk = GLPK();
    const tileScores = getTileScores();
    const tileProps  = getTileProperties();

    const rows = mapData.length;
    const cols = mapData[0].length;

    // -----------------------------------------------------------------------
    // Classify tile positions
    // "graphTiles" includes all non-blocking tiles PLUS fillable tiles.
    // -----------------------------------------------------------------------
    const graphTiles    = [];   // All tiles that participate in the MILP graph
    const wallPlaceable = [];   // Tiles where walls can be placed
    const fillableSet   = new Set(); // fillable = blocksMovement AND wallPlaceable
    const tileSet       = new Set(); // fast membership test for adjacency checks
    const tileScoreMap  = {};        // key → score when tile is in pen
    const boundary      = new Set();
    let   home          = null;

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            const tileName = mapData[r][c];
            const props    = tileProps[tileName] || {};
            const blocks   = props.blocksMovement  || false;
            const placeable = props.wallPlaceable  || false;
            const isFillable = blocks && placeable;

            // Skip purely blocking tiles (water/wall) that cannot be filled
            if (blocks && !placeable) continue;

            const key = `${r},${c}`;
            graphTiles.push([r, c]);
            tileSet.add(key);

            if (isFillable) {
                fillableSet.add(key);
                // Score when filled: use wallTransformsTo tile's score
                const transformed = props.wallTransformsTo;
                tileScoreMap[key] = transformed
                    ? (tileScores[transformed] ?? 1)
                    : 1;
            } else {
                tileScoreMap[key] = tileScores[tileName] ?? 1;
            }

            if (placeable) wallPlaceable.push([r, c]);
            if (tileName === 'home') home = [r, c];
            if (r === 0 || r === rows - 1 || c === 0 || c === cols - 1) {
                boundary.add(key);
            }
        }
    }

    if (!home) {
        return {
            goalArea: 0, optimalWallCount: 0, optimalSolution: [], feasible: false,
            error: 'No home tile found',
        };
    }

    const homeKey = `${home[0]},${home[1]}`;
    if (boundary.has(homeKey)) {
        return {
            goalArea: 0, optimalWallCount: 0, optimalSolution: [], feasible: false,
            error: 'Home is on boundary - cannot pen',
        };
    }

    // -----------------------------------------------------------------------
    // Build directed edges between adjacent graph tiles
    // -----------------------------------------------------------------------
    const DIRS = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    const edges     = [];
    const outEdges  = {};
    const inEdges   = {};

    for (const [r, c] of graphTiles) {
        outEdges[`${r},${c}`] = [];
        inEdges[`${r},${c}`]  = [];
    }
    for (const [r, c] of graphTiles) {
        for (const [dr, dc] of DIRS) {
            const nr = r + dr, nc = c + dc;
            const nk = `${nr},${nc}`;
            if (tileSet.has(nk)) {
                edges.push([[r, c], [nr, nc]]);
                outEdges[`${r},${c}`].push([nr, nc]);
                inEdges[nk].push([r, c]);
            }
        }
    }

    const n            = graphTiles.length;
    const placeableSet = new Set(wallPlaceable.map(([r, c]) => `${r},${c}`));

    // Variable name helpers
    const sVar = ([r, c])                       => `s_${r}_${c}`;
    const wVar = ([r, c])                       => `w_${r}_${c}`;
    const fVar = ([r1, c1], [r2, c2])           => `f_${r1}_${c1}_${r2}_${c2}`;

    // -----------------------------------------------------------------------
    // Build GLPK problem
    // -----------------------------------------------------------------------
    const subjectTo = [];
    const binaries  = [];

    // Declare binary variables
    for (const tile of graphTiles)    binaries.push(sVar(tile));
    for (const tile of wallPlaceable) binaries.push(wVar(tile));

    // --- Constraint 1: home is in the pen -----------------------------------
    subjectTo.push({
        name: 'home_in_pen',
        vars: [{ name: sVar(home), coef: 1 }],
        bnds: { type: glpk.GLP_FX, lb: 1, ub: 1 },
    });

    // --- Constraint 2: boundary tiles are NOT in the pen -------------------
    for (const [r, c] of graphTiles) {
        const key = `${r},${c}`;
        if (boundary.has(key)) {
            subjectTo.push({
                name: `bound_${r}_${c}`,
                vars: [{ name: sVar([r, c]), coef: 1 }],
                bnds: { type: glpk.GLP_FX, lb: 0, ub: 0 },
            });
        }
    }

    // --- Constraint 3a: normal wall-placeable tiles — wall excludes from pen
    // s[tile] + w[tile] <= 1
    for (const tile of wallPlaceable) {
        const key = `${tile[0]},${tile[1]}`;
        if (!fillableSet.has(key)) {
            subjectTo.push({
                name: `wall_excl_${tile[0]}_${tile[1]}`,
                vars: [
                    { name: sVar(tile), coef: 1 },
                    { name: wVar(tile), coef: 1 },
                ],
                bnds: { type: glpk.GLP_UP, lb: 0, ub: 1 },
            });
        }
    }

    // --- Constraint 3b: fillable tiles — only in pen when filled (wall placed)
    // s[tile] <= w[tile]  →  s[tile] - w[tile] <= 0
    for (const [r, c] of graphTiles) {
        const key = `${r},${c}`;
        if (fillableSet.has(key)) {
            subjectTo.push({
                name: `fill_${r}_${c}`,
                vars: [
                    { name: sVar([r, c]), coef:  1 },
                    { name: wVar([r, c]), coef: -1 },
                ],
                bnds: { type: glpk.GLP_UP, lb: 0, ub: 0 },
            });
        }
    }

    // --- Constraint 4: wall budget -----------------------------------------
    if (wallPlaceable.length > 0) {
        subjectTo.push({
            name: 'wall_budget',
            vars: wallPlaceable.map(t => ({ name: wVar(t), coef: 1 })),
            bnds: { type: glpk.GLP_UP, lb: 0, ub: maxWalls },
        });
    }

    // --- Constraint 5: vertex cut ------------------------------------------
    // s[i] - s[j] <= barrier_i + barrier_j
    // where:
    //   barrier_x = w[x]     for normal placeable  (coef -1 on left, RHS +0)
    //   barrier_x = 1 - w[x] for fillable           (coef +1 on left, RHS +1)
    //   barrier_x = 0        for non-placeable
    for (const [[ri, ci], [rj, cj]] of edges) {
        const ki = `${ri},${ci}`;
        const kj = `${rj},${cj}`;
        const vars = [
            { name: sVar([ri, ci]), coef:  1 },
            { name: sVar([rj, cj]), coef: -1 },
        ];
        let ub = 0;

        if (placeableSet.has(ki)) {
            if (fillableSet.has(ki)) {
                vars.push({ name: wVar([ri, ci]), coef: 1 });
                ub += 1;
            } else {
                vars.push({ name: wVar([ri, ci]), coef: -1 });
            }
        }
        if (placeableSet.has(kj)) {
            if (fillableSet.has(kj)) {
                vars.push({ name: wVar([rj, cj]), coef: 1 });
                ub += 1;
            } else {
                vars.push({ name: wVar([rj, cj]), coef: -1 });
            }
        }

        subjectTo.push({
            name: `vcut_${ri}_${ci}_${rj}_${cj}`,
            vars,
            bnds: { type: glpk.GLP_UP, lb: 0, ub },
        });
    }

    // --- Constraint 6: flow conservation (pen connectivity via home) --------
    // Home (source): out_flow - in_flow == sum(s[t] for t != home)
    {
        const vars = [];
        for (const [nr, nc] of outEdges[homeKey]) {
            vars.push({ name: fVar(home, [nr, nc]), coef: 1 });
        }
        for (const [nr, nc] of inEdges[homeKey]) {
            vars.push({ name: fVar([nr, nc], home), coef: -1 });
        }
        for (const [r, c] of graphTiles) {
            if (r !== home[0] || c !== home[1]) {
                vars.push({ name: sVar([r, c]), coef: -1 });
            }
        }
        subjectTo.push({
            name: 'flow_home',
            vars,
            bnds: { type: glpk.GLP_FX, lb: 0, ub: 0 },
        });
    }

    // Non-home tiles: in_flow - out_flow == s[tile]
    for (const [r, c] of graphTiles) {
        if (r === home[0] && c === home[1]) continue;
        const key  = `${r},${c}`;
        const vars = [];
        for (const [nr, nc] of inEdges[key]) {
            vars.push({ name: fVar([nr, nc], [r, c]), coef: 1 });
        }
        for (const [nr, nc] of outEdges[key]) {
            vars.push({ name: fVar([r, c], [nr, nc]), coef: -1 });
        }
        vars.push({ name: sVar([r, c]), coef: -1 });
        subjectTo.push({
            name: `flow_${r}_${c}`,
            vars,
            bnds: { type: glpk.GLP_FX, lb: 0, ub: 0 },
        });
    }

    // --- Constraint 7: flow capacity — flow only through pen tiles ----------
    // f[(i,j)] <= n * s[i]  and  f[(i,j)] <= n * s[j]
    for (const [[ri, ci], [rj, cj]] of edges) {
        const fv = fVar([ri, ci], [rj, cj]);
        subjectTo.push({
            name: `fcap_i_${ri}_${ci}_${rj}_${cj}`,
            vars: [
                { name: fv,                coef:  1 },
                { name: sVar([ri, ci]),    coef: -n },
            ],
            bnds: { type: glpk.GLP_UP, lb: 0, ub: 0 },
        });
        subjectTo.push({
            name: `fcap_j_${ri}_${ci}_${rj}_${cj}`,
            vars: [
                { name: fv,                coef:  1 },
                { name: sVar([rj, cj]),    coef: -n },
            ],
            bnds: { type: glpk.GLP_UP, lb: 0, ub: 0 },
        });
    }

    // -----------------------------------------------------------------------
    // Objective: maximise enclosed area with tiny wall penalty as tiebreaker
    // -----------------------------------------------------------------------
    const objVars = [];
    for (const [r, c] of graphTiles) {
        const key = `${r},${c}`;
        objVars.push({ name: sVar([r, c]), coef: tileScoreMap[key] ?? 1 });
    }
    for (const tile of wallPlaceable) {
        objVars.push({ name: wVar(tile), coef: -0.0001 });
    }

    const lp = {
        name: 'PenThePet',
        objective: {
            direction: glpk.GLP_MAX,
            name: 'obj',
            vars: objVars,
        },
        subjectTo,
        binaries,
    };

    // -----------------------------------------------------------------------
    // Solve (120 s time limit, silent output)
    // -----------------------------------------------------------------------
    const opts = { msglev: glpk.GLP_MSG_OFF, tmlim: 120 };
    const result = glpk.solve(lp, opts);

    if (result.result.status !== glpk.GLP_OPT) {
        return {
            goalArea: 0, optimalWallCount: 0, optimalSolution: [], feasible: false,
            error: `Solver status: ${result.result.status}`,
        };
    }

    const goalArea = Math.round(result.result.z);
    const wallPositions = [];
    for (const tile of wallPlaceable) {
        const val = result.result.vars[wVar(tile)];
        if (val !== undefined && val > 0.5) {
            wallPositions.push([tile[0], tile[1]]);
        }
    }

    return {
        goalArea,
        optimalWallCount: wallPositions.length,
        optimalSolution: wallPositions,
        feasible: true,
    };
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

function main() {
    const fs   = require('fs');
    const args = process.argv.slice(2);

    let fileArg = null;
    for (let i = 0; i < args.length; i++) {
        if ((args[i] === '--file' || args[i] === '-f') && i + 1 < args.length) {
            fileArg = args[++i];
        }
    }

    let inputData;
    try {
        const raw = fileArg
            ? fs.readFileSync(fileArg, 'utf-8')
            : fs.readFileSync('/dev/stdin', 'utf-8');
        inputData = JSON.parse(raw);
    } catch (err) {
        process.stderr.write(`Failed to read input: ${err.message}\n`);
        process.exit(1);
    }

    const result = solveMap(inputData.map, inputData.maxWalls);
    process.stdout.write(JSON.stringify(result) + '\n');
}

module.exports = { solveMap };

if (require.main === module) {
    main();
}
